apps/enc.c

```diff
@@ -89,7 +93,8 @@ const OPTIONS enc_options[] = {
     {"iv", OPT_IV, 's', "IV in hex"},
     {"md", OPT_MD, 's', "Use specified digest to create a key from the passphrase"},
     {"iter", OPT_ITER, 'p', "Specify the iteration count and force use of PBKDF2"},
-    {"pbkdf2", OPT_PBKDF2, '-', "Use password-based key derivation function 2"},
+    {"pbkdf2", OPT_PBKDF2, '-', "Use password-based key derivation function 2 with"},
+    {OPT_MORE_STR, 0, 0, "default iteration count of " STR(PBKDF2_ITER_DEFAULT)},
```

Comment:
This looks misplaced, I would rather do this:

``` C
    {"iter", OPT_ITER, 'p', "Specify the iteration count and force use of PBKDF2"},
    {OPT_MORE_STR, 0, 0, "Default: " STR(PBKDF2_ITER_DEFAULT)},
    {"pbkdf2", OPT_PBKDF2, '-', "Use password-based key derivation function 2"},
```

Reply:
Maybe I misunderstood @levitte's point - but isn't the point that this extra text should be against the "-iter" option, rather than the "-pbkdf2" option?

Reply:
@mattcaswell understood what I'm suggesting

Reply:
The code revolving around `iter` is very confusing, BTW...  it suggests that `iter = 1` is a default too, but that's actually never used as far as I can tell.

Reply:
My thought was that there is no default if you specify `-iter`, you've specified the number of iterations directly.

The default **only** applies if you use `-pbkdf2` and not `-iter`, so that's where I put it.

This makes sense to me.

Reply:
Like I said, the code is confusing.  A slight bit of refactoring would make it less so:

``` C
diff --git a/apps/enc.c b/apps/enc.c
index 3a55d72c7a..b46808c7bd 100644
--- a/apps/enc.c
+++ b/apps/enc.c
@@ -125,7 +125,7 @@ int enc_main(int argc, char **argv)
     unsigned char key[EVP_MAX_KEY_LENGTH], iv[EVP_MAX_IV_LENGTH];
     unsigned char *buff = NULL, salt[PKCS5_SALT_LEN];
     int pbkdf2 = 0;
-    int iter = 0;
+    int iter = PBKDF2_ITER_DEFAULT;
     long n;
     int streamable = 1;
     int wrap = 0;
@@ -286,8 +286,6 @@ int enc_main(int argc, char **argv)
             break;
         case OPT_PBKDF2:
             pbkdf2 = 1;
-            if (iter == 0)    /* do not overwrite a chosen value */
-                iter = 10000;
             break;
         case OPT_NONE:
             cipher = NULL;
@@ -323,9 +321,6 @@ int enc_main(int argc, char **argv)
     if (dgst == NULL)
         dgst = (EVP_MD *)EVP_sha256();
 
-    if (iter == 0)
-        iter = 1;
-
     /* It must be large enough for a base64 encoded line */
     if (base64 && bsize < 80)
         bsize = 80;
```

And yes, this refactoring doesn't introduce any actual change, but it does clarify things a bit.

Reply:
There is a slight change - if you specify `-iter 0` the original will set it to 1, your refactoring keeps the 0.

Reply:
The issue here isn't about the code being confusing or not.  It's about what help is best for the user.
If a user needs to look at the code to figure out what's going on, we've failed.

Refactoring the code will mean this can't go back to 3.1 and 3.0.

Reply:
I still find your variant confusing, @paulidale, for the reason stated in https://github.com/openssl/openssl/pull/20471#discussion_r1131063325

Reply:
We disagree.  Anyone else have a thought here?

Reply:
I would personally prefer to see the default documented against the "-iter" option. The pbkdf2 option could always reference the iter option. E.g. "Use password-based key derivation function 2. Use -iter to change the iteration count", and then -iter would be "Specify the iteration count and force use of PBKDF2. Default 10000"

Reply:
Reworded and adjusted.